receiver/m365receiver/client.go

```diff
@@ -0,0 +1,460 @@
+	resp, err := m.makeRequest(ctx, Sub, endpoint)
+	if err != nil {
+		return err
+	}
+	defer func() { _ = resp.Body.Close() }()
+
+	// troubleshoot error handling, mainly sub already enabled
+	// no error if sub already enabled, not troubleshooting stale token
+	// only called while code is synchronous right after a GetToken call
+	// if token is stale, regenerating it won't fix anything
+	if resp.StatusCode != 200 {
+		if resp.StatusCode == 400 { // subscription already started possibly
+			respErr, err := m.readInErr(resp)
+			if err != nil {
+				return err
+			}
+			if respErr.Error.Message == "The subscription is already enabled. No property change." {
+				return nil
+			}
+		}
+		//unsure what the error is
+		return fmt.Errorf("got non 200 status code from request, got %d", resp.StatusCode)
+	}
+	//if StatusCode == 200, then subscription was successfully started
+	return nil
+}
+
+// followLink will follow the response of a first request that has a link to the actual content
+func (m *m365Client) followLink(ctx context.Context, endpoint string) ([]byte, error) {
+	resp, err := m.makeRequest(ctx, Default, endpoint)
+	if err != nil {
+		return []byte{}, err
+	}
+	defer func() { _ = resp.Body.Close() }()
+
+	// troubleshoot error code
+	if resp.StatusCode != 200 {
+		respErr, err := m.readInErr(resp)
+		if err != nil {
+			return []byte{}, err
+		}
+		if respErr.Message == "Authorization has been denied for this request." {
+			return []byte{}, fmt.Errorf("authorization denied")
+		}
+		return []byte{}, fmt.Errorf("got non 200 status code from request, got %d", resp.StatusCode)
+	}
+
+	return io.ReadAll(resp.Body)
+}
+
+// parseBody takes the byte[] response and parses it into string objects
+func (m *m365Client) parseBody(body []byte) []string {
+	data := strings.Split(string(body), "},{\"C")
+	last := len(data) - 1
+	data[0] = strings.TrimPrefix(data[0], "[{\"C")
+	data[last] = strings.TrimSuffix(data[last], "}]")
+	return data
+}
+
+// reads in errors for responses
+func (m *m365Client) readInErr(resp *http.Response) (apiError, error) {
+	var respErr apiError
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return apiError{}, err
+	}
+	err = json.Unmarshal(body, &respErr)
+	if err != nil {
+		return apiError{}, err
+	}
+	return respErr, nil
+}
+
+// reads in content uri for GetJSON
+func (m *m365Client) readInContent(resp *http.Response) (string, error) {
+	// check if contentUri field if present
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return "", err
+	}
+	if string(body) == "[]" { // if body is empty, no new log data available
+		return "", nil
+	}
+	// extract contentUri field
+	lr := []struct {
+		Content string `json:"contentUri"`
+	}{}
+	if err = json.Unmarshal(body, &lr); err != nil {
+		return "", err
+	}
+	return lr[0].Content, nil
+}
+
+// handles requests for client functions
+func (m *m365Client) makeRequest(ctx context.Context, r ReqType, endpoint string, opts ...Option) (*http.Response, error) {
```

Comment:
I think this function should be more generic, and contain less knowledge about the caller(s).

Something like the following, with a few new options for setting auth or providing a body. The body would then be turned into a strings.NewReader outside of the scope of the function where appropriate, and the callers would set the appropriate http method.

```
func (m *m365Client) makeRequest(ctx context.Context, httpMethod, endpoint string,  opts ...Option) (*http.Response, error) {

    req, err := http.NewRequestWithContext(ctx, "POST", endpoint, body)
    if err != nil {
        return nil, err
    }
    for _, o := range opts {
        o(req)
    }
    
    return m.client.Do(req)
}

func WithToken(token string) Option {
	return func(r *http.Request) {
		r.Header.Set("Authorization", "Bearer "+m.token)
	}
}

func WithBody(body io.ReadCloser) Option {
	return func(r *http.Request) {
		r.Body = body
		r.Header.Set("Content-Type", "application/x-www-form-urlencoded")
	}
}
```